feature: onedal verbose profiler (#3155)
uxlfoundation/oneDAL: docs/source/contribution/profiling.rst

.. Copyright contributors to the oneDAL project
..
.. Licensed under the Apache License, Version 2.0 (the "License");
.. you may not use this file except in compliance with the License.
.. You may obtain a copy of the License at
..
..     http://www.apache.org/licenses/LICENSE-2.0
..
.. Unless required by applicable law or agreed to in writing, software
.. distributed under the License is distributed on an "AS IS" BASIS,
.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.. See the License for the specific language governing permissions and
.. limitations under the License.

Using oneDAL Verbose Mode
=========================

When building applications that call Intel® oneAPI Data Analytics Library functions, it may be useful to determine:

- Which computational kernels are called
- What parameters are passed to them
- How much time is spent to execute the functions
- Which device (CPU/GPU) the kernel is executed on

You can get an application to print this information to a standard output
device by enabling **Intel® oneAPI Data Analytics Library Verbose Mode**.

> All logs are printed to std::cerr.

Reviewer: ```suggestion
All logs are printed to the `stderr` stream.
```

> When Verbose mode is active in oneDAL, every call of a verbose-enabled function finishes with 

Reviewer: This doesn't appear to be the case, so perhaps these docs could be updated. Otherwise, verbosity level 3 should print in python as it does indeed exit the function.

printing a human-readable line describing the call. Even if the application gets terminated during 
the function call, information for that function will be printed. 
The first call to a verbose-enabled function also prints a version information line.

For GPU applications, additional information (one or more GPU information lines) will also 
be printed on the first call to a verbose-enabled function, following the version information lines printed
for the host CPU.

Verbosity Levels
----------------

We have common implementations for verbose mode with CPU and GPU applications.

>   - Levels:

Reviewer: ```suggestion
 Levels:
 
```

Logger
------

**Logger** is a subtool used to capture information about each launched kernel.

Example output (for ``cov_dense_batch``):

::

    auto oneapi::dal::covariance::backend::compute_kernel_dense_impl<float>::operator()(const descriptor_t &, const parameters_t &, const input_t &)::(anonymous class)::operator()() const [Float = float]
    Profiler task_name: allreduce_sums, q_

    auto oneapi::dal::covariance::backend::compute_kernel_dense_impl<float>::operator()(const descriptor_t &, const parameters_t &, const input_t &)::(anonymous class)::operator()() const [Float = float]
    Profiler task_name: gemm, q_

Each entry shows the ``__PRETTY_FUNCTION__`` along with the corresponding ``PROFILER_TASK`` and its arguments.

The primary purpose of the Logger is debugging — especially useful for identifying the last called function before a segmentation fault or crash (if the function includes a profiler tag).

Tracer
------

**Tracer** is a subtool designed to collect and display the execution time of each ``profiler_task``.

Example output (for ``cov_dense_batch``):

::

>     reduction.reduce_by_columns        81.72ms

Reviewer: I think this log is outdated now that there's info from threaded tasks from the main thread.

Example output (for ``cov_dense_batch``):

::

    Algorithm tree analyzer
    |-- covariance_algo_compute               time: 501.60ms   100.00%
    |   |-- compute_sums                      time: 81.56ms    16.26%
    |   |   |-- reduction.reduce_by_columns   time: 81.09ms    16.17%
    |   |-- allreduce_sums                   time: 8.78us      0.00%
    |   |-- gemm                             time: 215.34ms    42.93%
    |   |   |-- blas.gemm                     time: 215.34ms   42.93%
    |   |-- allreduce_xtx                    time: 6.85us      0.00%
    |   |-- allreduce_rows_count_global      time: 422ns       0.00%
    |   |-- compute_correlation              time: 154.44ms    30.79%
    |   |-- compute_means                    time: 50.19ms     10.01%
    |---(end)

    ONEDAL KERNEL_PROFILER: kernels total time 501.60ms

The Analyzer helps visualize nested kernel calls and their contribution to the total runtime. It's especially helpful for understanding which tasks call which kernels and how much time each part consumes.

General Notes on Verbose Mode
-----------------------------

> 

Reviewer: Here could mention what are the requirements for triggering the analyzer, which doesn't show up on python.

.. Copyright contributors to the oneDAL project
..
.. Licensed under the Apache License, Version 2.0 (the "License");
.. you may not use this file except in compliance with the License.
.. You may obtain a copy of the License at
..
..     http://www.apache.org/licenses/LICENSE-2.0
..
.. Unless required by applicable law or agreed to in writing, software
.. distributed under the License is distributed on an "AS IS" BASIS,
.. WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.. See the License for the specific language governing permissions and
.. limitations under the License.

Using oneDAL Verbose Mode
=========================

When building applications that call Intel® oneAPI Data Analytics Library functions, it may be useful to determine:

> - which computational functions are called
> - what parameters are passed to them
> - how much time is spent to execute the functions
> - (for GPU applications) which GPU device the kernel is executed on

Reviewer: ```suggestion
- Which computational kernels are called
- What parameters are passed to them
- How much time is spent to execute the functions
- Which device (CPU/GPU) the kernel is executed on
```

> You can get an application to print this information to a standard output

Reviewer: Worth mentioning that some logs are going to `stdout` and some to `stderr`.

> device by enabling **Intel® oneAPI Data Analytics Library Verbose**.

Reviewer: ```suggestion
device by enabling **Intel® oneAPI Data Analytics Library Verbose Mode**.
```

When Verbose mode is active in oneDAL, every call of a verbose-enabled function finishes with 

> printing a human-readable line describing the call. Even, if your application gets terminated during 

Reviewer: ```suggestion
printing a human-readable line describing the call. Even if the application gets terminated during 
```

the function call, information for that function will be printed. 
The first call to a verbose-enabled function also prints a version information line.

For GPU applications, additional information (one or more GPU information lines) will also 

> be printed by the first call to a verbose-enabled function, following the version information line printed

Reviewer: ```suggestion
be printed on the first call to a verbose-enabled function, following the version information lines printed
```

for the host CPU.

> Modes of Verbose

Reviewer: ```suggestion
Verbosity Levels
```

----------------

> We have common implementations for verbose with CPU and GPU applications.

Reviewer: ```suggestion
We have common implementations for verbose mode with CPU and GPU applications.
```

>   - **Levels**:

Reviewer: ```suggestion
Levels:
```

> - **Applications**:

Reviewer: ```suggestion
- **Levels**:
```

  - Disabled (default)
  - Logger enabled
  - Tracer enabled

>   - Analyzer enabled

Reviewer: Could be useful to explain what this is and show an example output snippet.

Setting Verbose Mode
--------------------

You can change the verbose mode either by setting the environment variable ``ONEDAL_VERBOSE``.

**Using environment variable:**

+--------------------------+-----------------------------------------------+
| Set ``ONEDAL_VERBOSE=0`` | Disabled (default)                            |
+--------------------------+-----------------------------------------------+
| Set ``ONEDAL_VERBOSE=1`` | Logger enabled                                |
+--------------------------+-----------------------------------------------+
| Set ``ONEDAL_VERBOSE=2`` | Tracer enabled                                |
+--------------------------+-----------------------------------------------+
| Set ``ONEDAL_VERBOSE=3`` | Analyzer enabled                              |
+--------------------------+-----------------------------------------------+
| Set ``ONEDAL_VERBOSE=4`` | All tools enabled (Logger + Tracer + Analyzer)|
+--------------------------+-----------------------------------------------+
| Set ``ONEDAL_VERBOSE=5`` | All tools and service functions enabled       |
+--------------------------+-----------------------------------------------+

> When enabled with timing, execution timing is taken synchronously, meaning previous kernel execution will block later kernels.

Reviewer: Is "enabled with timing" the same as "tracer enabled"?

General Notes on Verbose Mode
-----------------------------

> 

Reviewer: Here could add a note that the environment variable only has an effect when set before oneDAL is loaded.

Author: It has been fixed and theoretically not an issue
